doc/rfc-ephemeral-disk-service.md

```diff
@@ -0,0 +1,482 @@
+```
+
+**Caveats:**
+
+* Breaks scenario (b) when only SCSI resource disk is available.
+
+### Long-term Plan: First-class Configuration in cloud-init
+
+We will independently work on a toggle for cloud-init to disable SCSI resource disk handling and key off this in a future version.
+
+Example:
+
+```yaml
+datasource:
+  Azure:
+    disable_resource_disk_handling: true
+```
+
+## Managing /etc/fstab
+
+An entry will be added to `/etc/fstab`.
+
+### fs_spec (source filesystem block device, uuid, etc.) 
+
+This will depend on the case:
+
+* aggregated units: `/dev/md/${AZURE_EPHEMERAL_DISK_SETUP_MDADM_NAME}_0`
+* single local NVMe: `/dev/disk/azure/local/by-serial/<serial>`
+* SCSI resource disk: `/dev/disk/azure/resource`
+
+### fs_vfstype (filesystem type)
+
+Matches `$AZURE_EPHEMERAL_DISK_SETUP_FS_TYPE`.
+
+### fs_mntops (mount options)
+
+Universal options: `defaults,nofail,comment=azure-ephemeral-disk-setup`
+
+* `defaults` to get the default options.
+* `nofail` to ensure lack of disk doesn't break boot process. 
+
+  * This allows VM to boot properly when:
+
+    * custom images contain artifacts from previous instance
+    * VM is re-allocated/migrated
+    * VM is migrated to new VM size, etc.
+
+* `comment=azure-ephemeral-disk-setup` to signal that this service manages this line.
+
+  * If there is any inconsistency in `/etc/fstab` lines with `comment=azure-ephemeral-disk-setup`, they will be deleted as part of setup.
+
+### fs_freq (dump)
+
+Always `0` (never dump).
+
+### fs_passno (fsck)
+
+Always `2` (check after root device).
+
+### Example `/etc/fstab` Entry
+
+```
+/dev/md/azure-ephemeral-md_0  /mnt  xfs  defaults,nofail,comment=azure-ephemeral-disk-setup  0  2
+```
+
+### Avoiding Data Loss and Empty Disk Checks
+
+If `/etc/fstab` does not contain an entry for the mount point, it is expected that the disks are empty.
+
+But on re-allocation, etc. we may have to deal with fresh disks that aren't already formatted. To ensure we avoid data loss, we will take the following precautions:
+
+* each local disk will be checked if in use with `findmnt <device>`
+
+  * This service runs after `local-fs.target`, so these disks should be actively mounted if in use by another mechanism.
+
+* `mdadm --create ...` and `mkfs.<fs_type>` will be invoked without `--force` flags to ensure they fail if the disks are partitioned/formatted or in-use
+
+If any of these checks or tools fail, the failure will be reported and service will exit with failure.
+
+In the worst case if there are cases where these tools don't detect in-use disks due to unforseen conditions, the actual risk to a user is quite low as these disks are expected to be ephemeral and this process occcurs early enough in boot to avoid negatively impacting running services in the worst case.
+
+The user must assume responsibility if they are using the disks in an unexpected manner that is impossible to detect, e.g. doing direct block read/writes without a filesystem mounted.
```

Comment:
I'm a little uncomfortable with this statement. In my experience, the "expected" behavior is that I, as the VM owner, own the disks, their provisioning, and their content. Admins have not historically needed to suppress system services in order to take ownership of their disks.

Reply:
Fair point! How about:
```
Given the limits on what we can reasonably detect, documentation will advise users not to opt-in to this service if using disks in a manner that would conflict and go undetected, e.g. doing direct block read/writes without a filesystem mounted.
```

Reply:
Whoops, I realize I did lose that blurb somewhere along the way,  I updated it to be really clear up front.

This is definitely going to be disabled by default.